### ISSUE TITLE
super('__construct') wierd behavior

### ISSUE THREAD

**Issue by simonklb** [haroldiedema/joii#10]
I get the following output when running the code below:
```
A test
Error: Method "setText" does not exist in the parent class. (called using 'super()')
```

```javascript
var A = Class({
    'public string text': 'nothing',
    __construct: function() {
        this.setText('test');
    },
});

var B = Class({extends: A}, {
    __construct: function() {
        this.super('__construct');
    }
});

var a = new A();
console.log('A', a.getText());

var b = new B();
console.log('B', b.getText());
```

Am I doing something wrong here or is it a bug in JOII?

**Comment by simonklb** [comment]
This might also be related:
```javascript
var A = Class({
    'public string text': 'Hello!',
    'public function printText': function() {
        console.log(this.getText())
    }
});

var B = Class({extends: A});

var a = new A();
var b = new B();

console.log('A');
a.printText();
console.log('B');
b.printText();
```
```
A
Hello!
B

b.printText();
  ^
TypeError: Object #<Object> has no method 'printText'
```

**Comment by haroldiedema** [comment]
Hi Simon,

Thanks for the report. The first one is definitely a bug. I've personally came across this a couple of times in some of my personal projects, but never found a way to crush this once and for all without breaking other functionality.

The problem occurs because of the way class definitions are linked together in the inheritance chain. Instead of simply copy-pasting functions, each function will have a generated function for itself that invokes its parent using the `super`-function, as seen here: https://github.com/haroldiedema/joii/blob/master/src/PrototypeBuilder.js#L265

Now, in your example, you're calling `super` yourself which then calls the parent constructor, which then calls the method that throws the error. It's extremely hard to keep track of which scope a method is called in, especially when using parent/child relationships due to the non-dereferencing nature of Object-types in JavaScript.

If you can find a way to solve this without breaking other functionality, please send me a pull request. I would _greatly_ appreciate it!


As for your second example; that does sound really strange. That should work just fine. I'll look into that.

**Comment by simonklb** [comment]
Hi again Harold!

I've started to look into it a little bit and found that this also currently does not work:

```
var A = Class({
    'public string text': 'Hello!',
    'public function printText': function() {
        console.log(this.text)
    }
});

var B = Class({extends: A});

var a = new A();
var b = new B();

console.log('A');
a.printText();
b.setText('Hello world!');
console.log('A2');
a.printText();
```

Output:
```
A
Hello!
A2
Hello!
```

This might perhaps give you some clues. Either way I'll continue to work on it. I found it quite interesting. :)

The second issue occurred because of the B class was missing a body and the parser expects a single object to be the body rather than parameters. I hotfixed it for myself inside the parser and it should be fairly easy to create a real solution for it as well.

On the other hand there isn't really any reason to extend a class without having anything more on top of it, so this might not even be something you want implemented. If so, perhaps there should be some exception for this scenario so that others doesn't run into the same trap. Let me know what do you think!

**Comment by simonklb** [comment]
Hi again Harold!

I've started to look into it a little bit and I found the problem quite interesting!

The second issue occurred because of the B class was missing a body and the parser expects a single object to be the body rather than parameters. I hotfixed it for myself inside the parser and it should be fairly easy to create a real solution for it as well.

On the other hand there isn't really any reason to extend a class without having anything more on top of it, so this might not even be something you want implemented. If so, perhaps there should be some exception for this scenario so that others doesn't run into the same trap. Let me know what do you think!

**Comment by haroldiedema** [comment]
Aahh, how silly that I didn't see this before...

```javascript
var B = Class({extends: A});
```

Feel free to send a pull request for this little bugger :)
I'm quite curious how you'd go about detecting a parameter-body or a class-body :smile: 

Thanks for your findings!!

**Comment by haroldiedema** [comment]
Fixed in 3.0.7.
